### PR TITLE
Ctv 3359 uat automation

### PR DESCRIPTION
**Pull request by addisonm**
### JIRA
https://infillion.atlassian.net/browse/CTV-3359

### Description
Introduce jenkins UAT automation

### Testing
A few test deployments

### Commit Message Subject(s)
LIST_PROPOSED_COMMIT_SUBJECT_LINES

### Additional Context
OPTIONAL_ADDITIONAL_INFO

### Related PRs
OPTIONAL_ADD_PR_LINKS

### Checks
- [ ] Includes unit test coverage _(if applicable)_
- [ ] Includes functional test coverage _(at feature-level, if applicable)_
- [ ] Proposed commit messages follow [team conventions](https://github.com/socialvibe/adlabs-wiki/blob/develop/git/README.md#commits)

### CODE REVIEW

File: src/deploy/jenkins-uat.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+            "Content-Type": "application/json",
+            "token": "TAR-HTML5_UAT",
+        },
+        body: JSON.stringify({
+            "whakapapaBranchName": "develop",
+            "devicePlatforms": "Unknown,Tizen,LG,XboxOne,FireTV,AndroidTV,Vizio,Comcast",
+            "skylineBranchName": "develop",
+            "skylineS3BranchDirName": "develop",
+            "fromBuildJobUrl": process.env.TRAVIS_BUILD_WEB_URL,
+            "sourceRepo": process.env.TRAVIS_REPO_SLUG
```
Comment (addisonm) [inline src/deploy/jenkins-uat.js]: This will look something like `socialvibe/Skyline`

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "truex-shared",
-  "version": "1.6.13",
+  "version": "1.7.1",
```
Comment (addisonm) [inline package.json]: I haven't made the 1.7.1 tag yet, figure I'd wait a bit given the current state

File: src/deploy/jenkins-uat.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
+const fetch = require('node-fetch');
+
+module.exports = () => {
+    const jenkinsUrl = "https://ctv-jenkins-02.truex.com/generic-webhook-trigger/invoke";
```
Comment (imann24) [inline src/deploy/jenkins-uat.js]: Can we allow for a env variable overload here? 
```javascript
const jenkinsUrl = process.env.JENKINS_URL || "https://ctv-jenkins-02.truex.com/generic-webhook-trigger/invoke";
```

File: package.json (hunk 2 above)
Comment (imann24) [reply package.json]: Good call! And thanks for handling the bump up to `1.7`. We've been meaning to do that for a bit

File: src/deploy/jenkins-uat.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    return fetch(jenkinsUrl, {
+        method: "POST",
+        headers: {
+            "Content-Type": "application/json",
+            "token": "TAR-HTML5_UAT",
+        },
+        body: JSON.stringify({
+            "whakapapaBranchName": "develop",
```
Comment (imann24) [inline src/deploy/jenkins-uat.js]: These may be worth environment variable overrides too

File: src/deploy/jenkins-uat.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+            "devicePlatforms": "Unknown,Tizen,LG,XboxOne,FireTV,AndroidTV,Vizio,Comcast",
+            "skylineBranchName": "develop",
+            "skylineS3BranchDirName": "develop",
+            "fromBuildJobUrl": process.env.TRAVIS_BUILD_WEB_URL,
```
Comment (imann24) [inline src/deploy/jenkins-uat.js]: Nice touch!

**Review by imann24 — APPROVED** [review]
Looks good to me! Allowing more overrides may be handy

File: src/deploy/jenkins-uat.js (hunk 3 above)
Comment (rblaa) [reply src/deploy/jenkins-uat.js]: this is a good idea

File: src/deploy/jenkins-uat.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
+const fetch = require('node-fetch');
+
+module.exports = () => {
```
Comment (rblaa) [inline src/deploy/jenkins-uat.js]: can we have a named function instead, so it is really clear what this is for, e.g.
```
// Inform Jenkins server to start an automation test run for the current branch.
function invokeAutomationTests() {
...
}
module.exports = invokeAutomationTests;

File: src/deploy/jenkins-uat.js (hunk 4 above)
Comment (rblaa) [inline src/deploy/jenkins-uat.js]: this needs to be parameterized, probably make an arg to the function, since the host repo knows the travis settings, not this shared function as such

File: src/deploy/jenkins-uat.js (hunk 3 above)
Comment (addisonm) [reply src/deploy/jenkins-uat.js]: The overloads don't seem too useful right now, if this is only going to be executed by the travis automation (and currently only does the defaults anyway.  But I can understand the sentiment of making it more future proof.

I wouldn't want to use environment variables for this though, and instead make it parameterized.

File: src/deploy/jenkins-uat.js (hunk 1 above)
Comment (rblaa) [inline src/deploy/jenkins-uat.js]: we should figure out if we want to use env vars here, or make these function args
I am inclined to make them function args, so that we can test this in isolation
the call can supply the args from the env vars instead

File: src/deploy/jenkins-uat.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+        body: JSON.stringify({
+            "whakapapaBranchName": "develop",
+            "devicePlatforms": "Unknown,Tizen,LG,XboxOne,FireTV,AndroidTV,Vizio,Comcast",
+            "skylineBranchName": "develop",
+            "skylineS3BranchDirName": "develop",
+            "fromBuildJobUrl": process.env.TRAVIS_BUILD_WEB_URL,
+            "sourceRepo": process.env.TRAVIS_REPO_SLUG
+        })
+    }).catch(error => {
+        console.log(error);
```
Comment (rblaa) [inline src/deploy/jenkins-uat.js]: let's put a bit on context so that the error message is clear:
```
console.error('error invoking Jenkins automation at ' + jenkinsUrl + '\n. ' +error);

**Review by rblaa — CHANGES_REQUESTED** [review]

File: src/deploy/jenkins-uat.js (hunk 3 above)
Comment (imann24) [reply src/deploy/jenkins-uat.js]: For the Jenkins URL, it feels worth having a way to change it without requiring a code deploy (especially in this shared package given how many others it touches). Mainly based on the fact the Jenkins URL has changed in the past

File: src/deploy/jenkins-uat.js (hunk 1 above)
Comment (addisonm) [reply src/deploy/jenkins-uat.js]: I'll make it arguments since I'm doing it for other things.  The isolation testing is a good byproduct as well and even allows potential for manual invocation

File: package.json (hunk 2 above)
Comment (rblaa) [inline package.json]: @addisonm can you use the 1.7.3 version and tag going forward
i need to make another intervening truex-shared fix for automation that i hope to merge today
so i will be taking the 1.7.2 tag
you will then need to merge that in to your branch first so that the final tags are cumulative

**Review by rblaa — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,8 +1,11 @@
 # Changelog
 
-## v1.7.1
+## v1.7.3
```
Comment (rblaa) [inline CHANGELOG.md]: should be 1.7.4

File: src/deploy/invokeAutomationTests.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+        method: "POST",
+        headers: {
+            "Content-Type": "application/json",
+            "token": "TAR-HTML5_UAT",
+        },
+        body: JSON.stringify({
+            "devicePlatforms": devicePlatforms || "Unknown,Tizen,LG,XboxOne,FireTV,AndroidTV,Vizio,Comcast",
+            "whakapapaBranchName": whakapapaBranchName,
+            "skylineBranchName": skylineBranchName,
+            "skylineS3BranchDirName": skylineS3BranchDirName,
```
Comment (rblaa) [inline src/deploy/invokeAutomationTests.js]: we can default this if missing, to be computed from the skylineBranchName

File: src/deploy/invokeAutomationTests.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+    return fetch(url, {
+        method: "POST",
+        headers: {
+            "Content-Type": "application/json",
+            "token": "TAR-HTML5_UAT",
+        },
+        body: JSON.stringify({
+            "devicePlatforms": devicePlatforms || "Unknown,Tizen,LG,XboxOne,FireTV,AndroidTV,Vizio,Comcast",
+            "whakapapaBranchName": whakapapaBranchName,
```
Comment (rblaa) [inline src/deploy/invokeAutomationTests.js]: default to develop if missing
is this actually needed for the caller to know?

Comment (rblaa) [reply src/deploy/invokeAutomationTests.js]: can you file a ticket for wakapapa to support an "All" value so we don't have to maintain the known set here
e.g. what happens when we can start including PS4 and PS5?

**Review by rblaa — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 8 above)
Comment (addisonm) [reply CHANGELOG.md]: I'll need to do 1.7.5 with how pushed the tags to test it on travis.

File: src/deploy/invokeAutomationTests.js (hunk 9 above)
Comment (addisonm) [reply src/deploy/invokeAutomationTests.js]: This looked more appropriate to be based off the pull request rather than skyline branch.  Since this is for the test branches no?  I'll move the guard here instead of from Skyline

File: src/deploy/deployHelper.js [hunk 11]
```diff
@@ -0,0 +1,8 @@
+
+const getS3BucketPath = (branch) => {
+    return `branch-test/skyline/${branch.replace(/[^a-z0-9_\-]/gi, "_").toLowerCase()}`;
```
Comment (rblaa) [inline src/deploy/deployHelper.js]: for this to be in trues-shared, the repo needs to be an arg
branch-test is also quite specific, and context-dependent
do this instead, rename the function to getS3BranchFolderName, use the function syntax since it is easier to read:
```
function getS3BranchFolderName(branch) {
  return branch.replace(/[^a-z0-9_\-]/gi, "_").toLowerCase();
}

now if you actually need the full path, then you need the prefix to be passed in
basically it really needs to be the caller, since master or develop for Skyline is different than master or develop for C3, etc.

**Review by rblaa — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,5 +1,8 @@
 # Changelog
 
+## v1.7.10
```
Comment (imann24) [inline CHANGELOG.md:3]: Looks like we're missing an update to the version in the `package.json` file

**Comment by rblaa** [comment]
travis checks are failing

**Review by rblaa — APPROVED** [review]
should fix the travis failures

**Comment by addisonm** [comment]
Noting that there are intermittent travis failures.  One is known, the other (which failed for this PR is not known, but was fixed with a travis re-run)
